contracts/MultiPartyEscrow.sol

```diff
@@ -0,0 +1,270 @@
+    public
+    {
+        token = ERC20(_token);
+    }
+  
+    function deposit(uint256 value) 
+    public
+    returns(bool) 
+    {
+        require(token.transferFrom(msg.sender, this, value), "Unable to transfer token to the contract");
+        balances[msg.sender] += value;
+        return true;
+    }
+    
+    function withdraw(uint256 value)
+    public
+    returns(bool)
+    {
+        require(balances[msg.sender] >= value);
+        require(token.transfer(msg.sender, value));
+        balances[msg.sender] -= value;
+        return true;
+    }
+    
+    //open a channel, tokan should be already being deposit
+    function open_channel(address  recipient, uint256 value, uint256 expiration, uint256 replica_id) 
+    public
+    returns(bool) 
+    {
+        require(balances[msg.sender] >= value);
+        channels[next_channel_id] = PaymentChannel({
+            sender       : msg.sender,
+            recipient    : recipient,
+            value        : value,
+            replica_id   : replica_id,
+            nonce        : 0,
+            expiration   : expiration
+        });
+        balances[msg.sender] -= value;
+        emit event_channel_open(next_channel_id, msg.sender, recipient, replica_id);
+        next_channel_id += 1;
+        return true;
+    }
+    
+
+
+    function deposit_and_open_channel(address  recipient, uint256 value, uint256 expiration, uint256 replica_id)
+    public
+    returns(bool)
+    {
+        require(deposit(value));
+        require(open_channel(recipient, value, expiration, replica_id));
+        return true;
+    }
+
+    //open a channel from the recipient side. Sender should send the signed permission to open the channel
+    function open_channel_by_recipient(address  sender, uint256 value, uint256 expiration, uint256 replica_id, uint256 message_nonce, bytes memory signature) 
+    public
+    returns(bool) 
+    {
+        require(balances[sender] >= value);
+        require(isValidSignature_open_channel_replaysafe(msg.sender, value, expiration, replica_id, message_nonce, signature, sender));
+
+        channels[next_channel_id] = PaymentChannel({
+            sender       : sender,
+            recipient    : msg.sender,
+            value        : value,
+            replica_id   : replica_id,
+            nonce        : 0,
+            expiration   : expiration
+        });
+        balances[sender] -= value;
+
+        emit event_channel_open(next_channel_id, sender, msg.sender, replica_id);
+        next_channel_id += 1;
+        return true;
+    }
+ 
+    function _channel_sendback_and_reopen(uint256 channel_id)
+    private
+    {
+        PaymentChannel storage channel = channels[channel_id];
+        balances[channel.sender]      += channel.value; 
+        channel.value                  = 0;
+        channel.nonce                 += 1;
+        channel.expiration             = 0;
+    }
+
+    // the recipient can close the channel at any time by presenting a
```

Comment:
I propose adding comment to explain that this method should be called iff there is no any channel for sender/recipient pair. If there is one then it must be reused. This is cheaper than opening new channel.